### PR TITLE
js derive proc macro

### PR DESCRIPTION
**Pull request by jazzfool**

### CODE REVIEW

File: derive/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "fueros_derive"
+description = "Derive macros for Fueros"
+version = "0.1.0"
+authors = ["You <you@example.com>"]
```
Comment (aleokdev) [inline derive/Cargo.toml]: Can you include authors/delete the attribute otherwise?

File: derive/src/js_enum.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                        )
+                    })
+                    .collect(),
+            });
+        }
+
+        let out_ident = quote::format_ident!("Js{}", ident);
+        let out_fields = variants
+            .iter()
+            .fold(Vec::new(), |mut v, x| {
```
Comment (aleokdev) [inline derive/src/js_enum.rs]: `.map` + `.flatten` + `.collect`?

File: derive/src/js_enum.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            .map(|(ident, ty)| {
+                quote! {
+                    pub #ident: Option<#ty>
+                }
+            });
+
+        let from_match_cases = variants
+            .iter()
+            .map(|x| {
+                //let full_variant_ident = quote::format_ident!("{}::{}", ident, x.name);
```
Comment (aleokdev) [inline derive/src/js_enum.rs]: Please clean this up

File: derive/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+mod js_enum;
+
+use proc_macro::TokenStream;
+
+#[proc_macro_derive(JsEnum)]
+pub fn js_enum(input: TokenStream) -> TokenStream {
```
Comment (aleokdev) [inline derive/src/lib.rs:7]: Why is this required? Can't re-export?

**Review by aleokdev — CHANGES_REQUESTED** [review]

File: derive/src/lib.rs (hunk 4 above)
Comment (jazzfool) [reply derive/src/lib.rs:7]: no, proc_macros must be defined in the top-level of a crate :(

File: derive/src/js_enum.rs (hunk 3 above)
Comment (jazzfool) [reply derive/src/js_enum.rs]: right away sire

File: derive/src/js_enum.rs (hunk 2 above)
Comment (jazzfool) [reply derive/src/js_enum.rs]: how foolish of me

File: derive/Cargo.toml (hunk 1 above)
Comment (jazzfool) [reply derive/Cargo.toml]: yes master

File: derive/src/lib.rs (hunk 4 above)
Comment (aleokdev) [reply derive/src/lib.rs:7]: agh, sucks

File: derive/src/js_enum.rs [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+            name: String,
+            in_fields: Vec<syn::Ident>,
+            out_fields: Vec<(syn::Ident, syn::Type)>,
+        }
+
+        let mut variants = Vec::new();
+        let mut is_unnamed = false;
+
+        let ident = &ast.ident;
+        for variant in &enum_data.variants {
```
Comment (aleokdev) [inline derive/src/js_enum.rs]: this can be inlined with a `map` and `collect`, also consider extracting long expressions into nameable functions

File: derive/src/js_enum_impl.rs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+                let sig = &method.sig;
+                let fn_name = &sig.ident;
+                let is_mut = method.sig.inputs.iter().find_map(|arg| {
+                    if let syn::FnArg::Receiver(arg) = arg {
+                        Some(arg)
+                    } else {
+                        None
+                    }
+                });
+                if is_mut.is_none() {
```
Comment (aleokdev) [inline derive/src/js_enum_impl.rs]: You can safely delete this and add a `?` at the end of the `is_mut` expr instead (also, please, use a helper function for obtaining all of that)

File: src/lib.rs [hunk 7]
```diff
@@ -19,16 +18,114 @@ pub fn main_js() -> Result<(), JsValue> {
...
+
+    pub fn change_player(&mut self, new_player_idx: u32) {
+        if let Edge::Set { player_idx } = self {
+            *player_idx = new_player_idx;
+        }
+    }
+}
+
 #[wasm_bindgen()]
 pub fn test() -> JsValue {
```
Comment (aleokdev) [inline src/lib.rs:47]: Move these tests to `tests/`

File: tests/app.rs [hunk 8]
```diff
@@ -1,24 +1,23 @@
...
 }
 
-
 // This runs a unit test in the browser, so it can use browser APIs.
 #[wasm_bindgen_test]
 fn web_test() {
     assert_eq!(1, 1);
 }
 
+/*
```
Comment (aleokdev) [inline tests/app.rs]: Why did you comment this out? I'd prefer if you deleted it entirely, but why did you do it in the first place?

**Review by aleokdev — APPROVED** [review]
The whole proc_macro code needs refactoring, but it's ok

File: tests/app.rs (hunk 8 above)
Comment (jazzfool) [reply tests/app.rs]: weird compile errors
dunno what half of this autogenerated stuff is anyway, looks totally unnecessary

File: src/lib.rs (hunk 7 above)
Comment (jazzfool) [reply src/lib.rs:47]: right away

**Comment by jazzfool** [comment]
should be ready to go
